File: New folder/cdp-sdk-8fcfac348487f586b224b77074f52594ad44198b/examples/typescript/evm/swaps/createSwapQuote.ts
// Usage: pnpm tsx evm/swaps/createSwapQuote.ts

/**
 * This example demonstrates how to create a swap quote with full transaction data.
 * 
 * IMPORTANT: createSwapQuote signals a soft commitment to swap and may reserve
 * funds on-chain. As such, it is rate-limited more strictly than getSwapPrice
 * to prevent abuse. Use getSwapPrice for more frequent price checks.
 * 
 * Use createSwapQuote when you need:
 * - Complete transaction data ready for execution
 * - Precise swap parameters with high accuracy
 * - To inspect transaction details before execution
 * - Full control over the swap execution process
 * 
 * Key differences from getSwapPrice:
 * - createSwapQuote: Returns full transaction data ready for execution
 * - createSwapQuote: May reserve funds on-chain (represents a soft commitment to swap)
 * - createSwapQuote: More strictly rate-limited to prevent abuse
 * - getSwapPrice: Returns pricing data only, no transaction details
 * - getSwapPrice: Can be queried more frequently, but data may be less precise
 * 
 * createSwapQuote is ideal for:
 * - Actually executing swaps (required step before execution)
 * - Implementing approval flows (checking allowance issues)
 * - Custom transaction handling or batching
 * - When you need guaranteed execution parameters
 * 
 * The quote includes:
 * - Transaction data (to, data, value, gas)
 * - Permit2 signature requirements for ERC20 swaps
 * - Exact amounts with slippage protection
 * - Gas estimates and potential issues
 * 
 * Note: For the simplest swap experience, use account.swap() which handles
 * quote creation and execution in one call. Use createSwapQuote when you
 * need more control over the process.
 */

import { CdpClient } from "@coinbase/cdp-sdk";
import { 
  formatUnits, 
  parseUnits, 
  formatEther,
  type Address
} from "viem";
import "dotenv/config";

// Network configuration
const NETWORK = "base"; // Base mainnet

// Token definitions for the example (using Base mainnet token addresses)
const TOKENS = {
  WETH: {
    address: "0x4200000000000000000000000000000000000006",
    symbol: "WETH",
    decimals: 18,
    isNativeAsset: false
  },
  USDC: {
    address: "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913",
    symbol: "USDC",
    decimals: 6,
    isNativeAsset: false
  },
};

// Create a CDP client
const cdp = new CdpClient();

async function main() {
  console.log(`Note: This example is using ${NETWORK} network.`);

  // Get or create an account to use for the swap
  const ownerAccount = await cdp.evm.getOrCreateAccount({ name: "SwapAccount" });
  console.log(`\nUsing account: ${ownerAccount.address}`);

  try {
    // Define the tokens we're working with
    const fromToken = TOKENS.WETH;
    const toToken = TOKENS.USDC;
    
    // Set the amount we want to send
    const fromAmount = parseUnits("0.1", fromToken.decimals); // 0.1 WETH
    
    console.log(`\nCreating a swap quote for ${formatEther(fromAmount)} ${fromToken.symbol} to ${toToken.symbol}`);
    
    // Create the swap quote transaction using the CDP client method
    console.log("\nFetching swap quote...");
    const swapQuote = await cdp.evm.createSwapQuote({
      network: NETWORK,
      toToken: toToken.address as Address,
      fromToken: fromToken.address as Address,
      fromAmount,
      taker: ownerAccount.address,
      slippageBps: 100, // 1% slippage tolerance
    });
    
    // Log swap details
    logSwapInfo(swapQuote, fromToken, toToken);
    
    // Validate the swap for any issues
    validateSwap(swapQuote);
    
    console.log("\nSwap quote created successfully. To execute this swap, you would need to:");
    console.log("1. Ensure you have sufficient token allowance for Permit2 contract");
    console.log("2. Submit the swap transaction (either using the CDP SDK or viem - more examples in account.swap.ts, smartAccount.submitSwapOperation.ts, and sendSwapTransactionWithViem.ts)");
    console.log("3. Wait for transaction confirmation");
    
    // Show how to execute the swap using the new execute() method
    console.log("\nAlternatively, you can execute the swap directly using the quote's execute() method:");
    console.log("```typescript");
    console.log("// Execute the swap directly from the quote");
    console.log("// The swap will be executed using the signerAddress (in the case of smart accounts) or taker address (in the case of regular EOA accounts)");
    console.log("const result = await swapQuote.execute({");
    console.log("  idempotencyKey: 'unique-key' // optional");
    console.log("});");
    console.log(`// Transaction hash: \${result.transactionHash}`);
    console.log("");
    console.log("// Or execute without any options");
    console.log("const result = await swapQuote.execute();");
    console.log("```");
    
  } catch (error) {
    console.error("Error creating swap quote:", error);
  }
}

/**
 * Logs information about the swap
 * @param swapQuote - The swap transaction data
 * @param fromToken - The token being sent
 * @param toToken - The token being received
 */
function logSwapInfo(
  swapQuote: any,
  fromToken: typeof TOKENS.WETH,
  toToken: typeof TOKENS.USDC
): void {
  if (!swapQuote.liquidityAvailable) {
    return;
  }

  console.log("\nSwap Quote Details:");
  console.log("-------------------");
  console.log(`Receive Amount: ${formatUnits(BigInt(swapQuote.toAmount), toToken.decimals)} ${toToken.symbol}`);
  console.log(`Min Receive Amount: ${formatUnits(BigInt(swapQuote.minToAmount), toToken.decimals)} ${toToken.symbol}`);
  console.log(`Send Amount: ${formatUnits(BigInt(swapQuote.fromAmount), fromToken.decimals)} ${fromToken.symbol}`);
  
  // Calculate and display price ratios
  const fromAmountBigInt = BigInt(swapQuote.fromAmount);
  const toAmountBigInt = BigInt(swapQuote.toAmount);
  const minToAmountBigInt = BigInt(swapQuote.minToAmount);
  
  // Calculate exchange rate: How many toTokens per 1 fromToken
  const fromToToRate = Number(toAmountBigInt) / (10 ** toToken.decimals) * 
                       (10 ** fromToken.decimals) / Number(fromAmountBigInt);
  
  // Calculate minimum exchange rate with slippage applied
  const minFromToToRate = Number(minToAmountBigInt) / (10 ** toToken.decimals) * 
                         (10 ** fromToken.decimals) / Number(fromAmountBigInt);
  
  // Calculate maximum toToken to fromToken ratio with slippage
  const maxToToFromRate = Number(fromAmountBigInt) / (10 ** fromToken.decimals) *
                         (10 ** toToken.decimals) / Number(minToAmountBigInt);

  // Calculate exchange rate: How many fromTokens per 1 toToken
  const toToFromRate = Number(fromAmountBigInt) / (10 ** fromToken.decimals) *
                       (10 ** toToken.decimals) / Number(toAmountBigInt);
  
  console.log("\nToken Price Calculations:");
  console.log("------------------------");
  console.log(`1 ${fromToken.symbol} = ${fromToToRate.toFixed(toToken.decimals)} ${toToken.symbol}`);
  console.log(`1 ${toToken.symbol} = ${toToFromRate.toFixed(fromToken.decimals)} ${fromToken.symbol}`);
  
  // Calculate effective exchange rate with slippage applied
  console.log("\nWith Slippage Applied (Worst Case):");
  console.log("----------------------------------");
  console.log(`1 ${fromToken.symbol} = ${minFromToToRate.toFixed(toToken.decimals)} ${toToken.symbol} (minimum)`);
  console.log(`1 ${toToken.symbol} = ${maxToToFromRate.toFixed(fromToken.decimals)} ${fromToken.symbol} (maximum)`);
  console.log(`Maximum price impact: ${((fromToToRate - minFromToToRate) / fromToToRate * 100).toFixed(2)}%`);
  
  console.log("\nSuggested Gas Details:");
  console.log("----------------------------------");
  console.log(`Gas: ${swapQuote.transaction.gas}`);
  console.log(`Gas Price: ${swapQuote.transaction.gasPrice}`);
}

/**
 * Validates the swap for any issues
 * @param swapQuote - The swap transaction data
 * @returns true if swap is valid, false if there are issues
 */
function validateSwap(swapQuote: any): boolean {
  console.log("\nValidating Swap Quote:");
  console.log("---------------------");
  
  if (!swapQuote.liquidityAvailable) {
    console.log("❌ Insufficient liquidity available for this swap.");
    return false;
  } else {
    console.log("✅ Liquidity available");
  }
  
  if (swapQuote.issues.balance) {
    console.log("\n❌ Balance Issues:");
    console.log(`Current Balance: ${swapQuote.issues.balance.currentBalance}`);
    console.log(`Required Balance: ${swapQuote.issues.balance.requiredBalance}`);
    console.log(`Token: ${swapQuote.issues.balance.token}`);
    console.log("\nInsufficient balance. Please add funds to your account.");
    return false;
  } else {
    console.log("✅ Sufficient balance");
  }

  if (swapQuote.issues.simulationIncomplete) {
    console.log("⚠️ WARNING: Simulation incomplete. Transaction may fail.");
  } else {
    console.log("✅ Simulation complete");
  }
  
  return true;
}

main().catch(console.error); 

File: New folder/cdp-sdk-8fcfac348487f586b224b77074f52594ad44198b/examples/python/evm/swaps/create_swap_quote.py
# Usage: uv run python evm/swaps/create_swap_quote.py
#!/usr/bin/env python3
# Usage: uv run python evm/create_swap_quote.py

"""
Example: Create Swap Quote and Execute

This example demonstrates the two-step swap process:
1. Create a swap quote to inspect the details
2. Execute the swap if satisfied with the quote

This approach allows you to show users exactly what they'll receive
before they commit to the swap.
"""

import asyncio
from decimal import Decimal

from cdp import CdpClient

from dotenv import load_dotenv

load_dotenv()


async def main():
    """Create a swap quote and execute it."""
    async with CdpClient() as cdp:
        # Get or create an account
        print("Getting account...")
        account = await cdp.evm.get_or_create_account(name="swap-example")
        print(f"Account address: {account.address}\n")
        
        # Token addresses on Base
        USDC = "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913"
        WETH = "0x4200000000000000000000000000000000000006"
        
        # Step 1: Create a swap quote
        print("Step 1: Creating swap quote...")
        quote = await cdp.evm.create_swap_quote(
            from_token=USDC,
            to_token=WETH,
            from_amount="1000000",  # 1 USDC (6 decimals)
            network="base",
            taker=account.address,
            slippage_bps=100,  # 1% slippage
            signer_address=account.address  # This enables quote.execute()
        )
        
        # Check if liquidity is available
        if not quote.liquidity_available:
            print("\n❌ Swap unavailable: Insufficient liquidity")
            print("   Try a smaller amount or a different token pair")
            return
        
        # Display the quote details
        print("\n📊 Swap Quote Details:")
        print(f"   Quote ID: {quote.quote_id}")
        print(f"   Selling: {Decimal(quote.from_amount) / Decimal(10**6):.2f} USDC")
        print(f"   Expected output: {Decimal(quote.to_amount) / Decimal(10**18):.6f} WETH")
        print(f"   Minimum output: {Decimal(quote.min_to_amount) / Decimal(10**18):.6f} WETH")
        print(f"   Max slippage: 1%")
        
        # Calculate the exchange rate
        from_amount_decimal = Decimal(quote.from_amount) / Decimal(10**6)
        to_amount_decimal = Decimal(quote.to_amount) / Decimal(10**18)
        rate = to_amount_decimal / from_amount_decimal
        print(f"   Exchange rate: 1 USDC = {rate:.8f} WETH")
        
        # Check if Permit2 signature is required
        if quote.requires_signature:
            print(f"   ⚠️  Permit2 signature required")
        
        # Ask for confirmation (in real app, this would be a UI interaction)
        print("\n🔄 Ready to execute swap...")
        
        try:
            # Step 2: Execute the swap using the quote
            print("Executing swap...")
            tx_hash = await quote.execute()
            
            print(f"\n✅ Swap executed successfully!")
            print(f"   Transaction hash: {tx_hash}")
            print(f"   View on Basescan: https://basescan.org/tx/{tx_hash}")
            
        except Exception as e:
            print(f"\n❌ Error executing swap: {e}")


if __name__ == "__main__":
    asyncio.run(main()) 